### PR TITLE
Refactor to better handle failures and reduce battery drain

### PR DESCRIPTION
**Pull request by bdraco**
- There was no disconnect callback
- There was no overall timeout
- The cache was not cleared if a char was missing or service resolution failed (there are BleakError guards that may make this ineffective for some chars that might need to be removed, but I don't know which devices this is safe to do for)
- `asyncio.wait_for` creates a task and is not recommended for use because its changes cancellation semantics
- To reduce battery drain, once device info has been read, don't read it again except for the sw_version

Note the battery drain reduction change is only effective with this change in HA
```diff
diff --git a/homeassistant/components/airthings_ble/__init__.py b/homeassistant/components/airthings_ble/__init__.py
index 1d62442f14d..a26675cc310 100644
--- a/homeassistant/components/airthings_ble/__init__.py
+++ b/homeassistant/components/airthings_ble/__init__.py
@@ -40,11 +40,11 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             f"Could not find Airthings device with address {address}"
         )
 
+    airthings = AirthingsBluetoothDeviceData(_LOGGER, elevation, is_metric)
+
     async def _async_update_method() -> AirthingsDevice:
         """Get data from Airthings BLE."""
         ble_device = bluetooth.async_ble_device_from_address(hass, address)
-        airthings = AirthingsBluetoothDeviceData(_LOGGER, elevation, is_metric)
-
         try:
             data = await airthings.update_device(ble_device)  # type: ignore[arg-type]
         except Exception as err:
```

### CODE REVIEW

File: airthings_ble/parser.py [hunk 1]
```diff
@@ -565,25 +597,49 @@ async def _get_service_characteristics(
...
-                    await client.stop_notify(characteristic.uuid)
+                    await client.stop_notify(characteristic)
 
-        return device
+    def _handle_disconnect(
+        self, disconnect_future: asyncio.Future[bool], client: BleakClient
+    ) -> None:
+        """Handle disconnect from device."""
+        self.logger.debug("Disconnected from %s", client.address)
+        disconnect_future.set_result(True)
```
Comment (LaStrada) [inline airthings_ble/parser.py:615]: I sometimes get this error:
```
2024-01-19 00:42:19.338 ERROR (MainThread) [root] A message handler raised an exception: invalid state
Traceback (most recent call last):
  File "src/dbus_fast/message_bus.py", line 802, in dbus_fast.message_bus.BaseMessageBus._process_message
  File "/home/lastrada/airthings/core/venv/lib/python3.11/site-packages/bleak/backends/bluezdbus/manager.py", line 998, in _parse_msg
    watcher.on_connected_changed(new_connected)
  File "/home/lastrada/airthings/core/venv/lib/python3.11/site-packages/bleak/backends/bluezdbus/client.py", line 168, in on_connected_changed
    self._disconnected_callback()
  File "/home/lastrada/airthings/airthings-ble/airthings_ble/parser.py", line 615, in _handle_disconnect
    disconnect_future.set_result(True)
asyncio.exceptions.InvalidStateError: invalid state
```

Comment (bdraco) [reply airthings_ble/parser.py]: fixed in 400ef9b70ba58879c9f3d8586558dfc449020530

**Comment by LaStrada** [comment]
2 linter errors, but other than that this PR seems to be working. I've tested this the last few days and I have had zero errors (except my previous comment).

**Review by LaStrada — APPROVED** [review]
Great work, thanks! 🚀

**Pull request merged**